public/icvar.h

```diff
@@ -11,21 +11,254 @@
+	{}
+
+	bool IsValid( ) const;
+	void Invalidate( );
+
+	uint16_t GetConVarIndex() const { return m_convarIndex; }
+	uint32_t GetIndex() const { return m_handleIndex; }
+
+private:
+	uint16_t m_convarIndex;
+	uint16_t m_unknown1;
+	uint32_t m_handleIndex;
+};
+static_assert(sizeof(ConVarHandle) == 0x8, "ConVarHandle is of the wrong size!");
+
+static const ConVarHandle INVALID_CONVAR_HANDLE = ConVarHandle();
+
+inline bool ConVarHandle::IsValid() const
+{
+	return m_convarIndex != 0xFFFF;
+}
+
+inline void ConVarHandle::Invalidate()
+{
+	m_convarIndex = 0xFFFF;
+	m_unknown1 = 0xFFFF;
+	m_handleIndex = 0x0;
+}
+
+//-----------------------------------------------------------------------------
+// Purpose: Internal structure of ConVar objects
+//-----------------------------------------------------------------------------
+enum EConVarType : short
+{
+	EConVarType_Invalid = -1,
+	EConVarType_Bool,
+	EConVarType_Int16,
+	EConVarType_UInt16,
+	EConVarType_Int32,
+	EConVarType_UInt32,
+	EConVarType_Int64,
+	EConVarType_UInt64,
+	EConVarType_Float32,
+	EConVarType_Float64,
+	EConVarType_String,
+	EConVarType_Color,
+	EConVarType_Vector2,
+	EConVarType_Vector3,
+	EConVarType_Vector4,
+	EConVarType_Qangle,
+	EConVarType_MAX
+};
+
+template<typename T>
+constexpr EConVarType TranslateConVarType();
+
+template<> constexpr EConVarType TranslateConVarType<bool>( void )		{ return EConVarType_Bool; }
+template<> constexpr EConVarType TranslateConVarType<int16_t>( void )	{ return EConVarType_Int16; }
+template<> constexpr EConVarType TranslateConVarType<uint16_t>( void )	{ return EConVarType_UInt16; }
+template<> constexpr EConVarType TranslateConVarType<int32_t>( void )	{ return EConVarType_Int32; }
+template<> constexpr EConVarType TranslateConVarType<uint32_t>( void )	{ return EConVarType_UInt32; }
+template<> constexpr EConVarType TranslateConVarType<int64_t>( void )	{ return EConVarType_Int64; }
+template<> constexpr EConVarType TranslateConVarType<uint64_t>( void )	{ return EConVarType_UInt64; }
+template<> constexpr EConVarType TranslateConVarType<float>( void )		{ return EConVarType_Float32; }
+template<> constexpr EConVarType TranslateConVarType<double>( void )	{ return EConVarType_Float64; }
+template<> constexpr EConVarType TranslateConVarType<const char*>( void ){ return EConVarType_String; }
+template<> constexpr EConVarType TranslateConVarType<Color>( void )		{ return EConVarType_Color; }
+template<> constexpr EConVarType TranslateConVarType<Vector2D>( void )	{ return EConVarType_Vector2; }
+template<> constexpr EConVarType TranslateConVarType<Vector>( void )	{ return EConVarType_Vector3; }
+template<> constexpr EConVarType TranslateConVarType<Vector4D>( void )	{ return EConVarType_Vector4; }
+template<> constexpr EConVarType TranslateConVarType<QAngle>( void )	{ return EConVarType_Qangle; }
+template<> constexpr EConVarType TranslateConVarType<void*>( void )		{ return EConVarType_Invalid; }
+
+template<typename T>
+class IConVar

```

Comment:
I'd like to suggest this structure if it's possible to implement it this way:
IConVar (or something with I) pure abstract, not templated. The header file doesn't contain anything but this interface with forward declaration of all the used types.
CConVarBase inherited from IConVar, not templated, with all the general data and logic. 
CConVar<T> inherited from CConVarBase with only the type-specific data and logic. 
Put the C* to cconvar.h/cconvar.cpp if it is acceptable by the design.

Reply:
To not give an impression that this was ignored, you can review the current structure if you have anything to add (Since some of your points were addressed). For the file naming, it's better to not rename something that's was like that since forever.